Replace trackingPhase1PU70 workflows with CA seeding

makortel opened the pull request:
~~This PR switches the default phase1 pixel seeding to cellular automaton (tuning by @felicepantaleo). The workflows of `trackingPhase1PU70` are replaced with workflows of the previous seeding (`trackingPhase1QuadProp`, suggestions for better name are welcome) to be kept as a reference for some time.~~ This PR replaces the `trackingPhase1PU70` workflows with CA seeding (tuning by @felicepantaleo). The goal is to switch the default phase1 tracking to CA seeding once the duplicate rate increase is better understood (and fixed/mitigated), therefore the "CA seeding" is already implemented under `trackingPhase1` sub-era, and the default trackign under `trackingPhase1QuadProp` era (suggestions for better name are welcome). Full cleanup of `trackingPhase1PU70` will come later (after the pre2 dust settles).

Here are MTV plots for 1000 ttbar+35PU events in 900pre1 (using 810pre16 DIGI-RAW as an input) comparing the CA seeding to the default.
https://mkortela.web.cern.ch/mkortela/tracking/validation/CMSSW_9_0_0_pre1_caseeding/
Timing decreases by ~5 % for the full trackingOnly reco job (becomes ~10 % when excluding most expensive other-than-iterative-tracking modules, mostly validation).

~~The previous seeding is kept workflows 10024.[45] (full/trackingOnly), so there should be no changes between old 10024.[01] and new 10024.[45].~~ 

~~To minimize merge conflicts the developments are based on @ebrondol's #16858 (in the hope of getting this in pre2).~~

Tested in CMSSW_9_0_X_2016-11-30-2300 (rebased on top of CMSSW_9_0_X after IB CMSSW_9_0_X_2016-12-09-1100 and before the next). No changes expected in standard workflows (special 10024.[45] show the changes from `trackingPhase1PU70` to CA). ~~for phase0, phase1 changes are described above.~~ ~~and for phase2 there are changes because of #16858 (but no changes on top of that).~~

@rovere @VinInn @felicepantaleo @ebrondol @mtosi

makortel committed:
Remove unnecessary imports

makortel committed:
Minimal work to disable trackingPhase1PU70 workflows

Configurations will be cleaned up later.

makortel committed:
Add trackingPhase1QuadProp first as a copy of trackingPhase1

This will preserve the current phase1 tracking while I transform
trackingPhase1 to use CA pixel seeding.

makortel committed:
Switch trackingPhase1 to CA seeding

Tuning by Felice Pantaleo.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
A new Pull Request was created by  @makortel (Matti Kortelainen) for CMSSW_9_0_X.

It involves the following packages:

Configuration/Eras
Configuration/PyReleaseValidation
Configuration/StandardSequences
RecoPixelVertexing/PixelTriplets
RecoTracker/FinalTrackSelectors
RecoTracker/IterativeTracking
RecoTracker/TkSeedingLayers
Validation/RecoTrack


@dmitrijus, @cvuosalo, @franzoni, @fabozzi, @cmsbuild, @srimanob, @slava77, @vanbesien, @hengne, @davidlange6 can you please review it and eventually sign? Thanks.
@ghellwig, @felicepantaleo, @GiacomoSguazzoni, @rovere, @VinInn, @Martin-Grunewald, @mschrode, @wmtford, @gpetruc, @dgulhan this is something you requested to watch as well.
@slava77, @smuzaffar you are the release manager for this.

cms-bot commands are listed here cms-sw/cmssw#13028

makortel commented:
@felicepantaleo Could you double-check the CA parameters?

felicepantaleo commented:
@makortel looks good to me

cmsbuild commented:
+1
Tested at: a1bb85022571deaa701146cf6648219a3d8881ef
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16911/16862/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16911/16862/summary.html

The workflows 1003.0, 1001.0, 1000.0, 140.53, 136.731, 4.22 have different files in step1_dasquery.log than the ones found in the baseline. You may want to check and retrigger the tests if necessary. You can check it in the "files" directory in the results of the comparisons

slava77 commented:
if #16885 is merged on Friday, this will likely need a rebase

makortel commented:
Seems that the system is smart-enough to not claim a conflict here after #16885 merge (as the commits from #16858 are exactly the same as in #16885).

slava77 commented:
@makortel 
for diff and incremental testing purposes, could you please rebase to get rid of the other commits.

makortel commented:
Sure, is there preference to do it now or wait an IB?

slava77 commented:
do it now, I'll wait for IB to test

cmsbuild commented:
Pull request #16911 was updated. @dmitrijus, @cvuosalo, @franzoni, @fabozzi, @cmsbuild, @srimanob, @slava77, @vanbesien, @hengne, @davidlange6 can you please check and sign again.

makortel commented:
Rebased on the head of CMSSW_9_0_X branch (better wait the next IB for tests).

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented on Configuration/Eras/python/Era_Run2_2017_trackingQuadProp_cff.py:
```diff
@@ -0,0 +1,7 @@
+from Configuration.Eras.ModifierChain_run2_2017_core_cff import run2_2017_core
+from Configuration.Eras.Modifier_trackingPhase1_cff import trackingPhase1
+
+Run2_2017_trackingPhase1QuadProp = cms.ModifierChain(run2_2017_core, trackingPhase1)
```
is this a spurious file or is it buggy?

makortel replied:
Spurious file, thanks for catching.

slava77 commented on RecoTracker/IterativeTracking/python/DetachedQuadStep_cff.py:
```diff
@@ -441,6 +461,7 @@
                                 detachedQuadStepTracks*
                                 detachedQuadStepClassifier1*detachedQuadStepClassifier2*
                                 detachedQuadStep)
+trackingPhase1.toReplaceWith(DetachedQuadStep, DetachedQuadStep.copyAndExclude([detachedQuadStepHitTriplets]))
```
since this is now the default, better eventually migrate to DetachedQuadStep appropriate for CA setup

makortel replied:
You mean making the default in the file CA, and customize it to `PixelQuadrupletEDProducer` for the "previous phase1" and phase2? If yes, that kind of reorganization will be part of future cleanup PR (also for other quad steps).

slava77 replied:
yes.
OK for the future PR.

felicepantaleo commented on RecoTracker/IterativeTracking/python/LowPtTripletStep_cff.py:
```diff
@@ -55,7 +57,8 @@
     nSigmaZ = 4.0
 ))
-trackingPhase1.toModify(lowPtTripletStepTrackingRegions, RegionPSet = dict(ptMin = 0.35)) # FIXME: Phase1PU70 value, let's see if we can lower it to Run2 value (0.2)
+trackingPhase1.toModify(lowPtTripletStepTrackingRegions, RegionPSet = dict(ptMin = 0.2))
```
@slava77 you can see some tests that I ran a while ago (810pre12) with this configuration:
https://fpantale.web.cern.ch/fpantale/offline_tuning/final/plots_timing/iterations.pdf
Many things happened in the mean time, but you can get the order of magnitude of the effect on timings on iterations.

slava77 replied:
if blue is baseline, then it looks like lowpttriplet step time has reduced.
Was your test done with the ptMin = 0.2 ?

felicepantaleo replied:
yep, all the parameters are the same as in this pr.

felicepantaleo replied:
and the timing of the original (blue), was with the original parameters, so 0.35 in this case

makortel replied:
Felice's plots should have exactly the same CA configuration as in here.

makortel commented on RecoPixelVertexing/PixelTriplets/python/quadrupletseedmerging_cff.py:
```diff
@@ -62,6 +62,8 @@
+from Configuration.Eras.Modifier_trackingPhase1QuadProp_cff import trackingPhase1QuadProp
+trackingPhase1QuadProp.toModify(PixelSeedMergerQuadruplets, **_forPhase1)
 from Configuration.Eras.Modifier_trackingPhase1PU70_cff import trackingPhase1PU70
 trackingPhase1PU70.toModify(PixelSeedMergerQuadruplets, **_forPhase1)
```
I will remove `trackingPhase1PU70` as soon as we have pre2 and this PR merged (whichever happens last). The only reason I didn't include it here was to avoid possible merge conflicts with phase2 developments targeting pre2.

makortel committed:
Remove spurious file

cmsbuild commented:
Pull request #16911 was updated. @dmitrijus, @cvuosalo, @franzoni, @fabozzi, @cmsbuild, @srimanob, @slava77, @vanbesien, @hengne, @davidlange6 can you please check and sign again.

makortel commented:
Removed `Configuration/Eras/python/Era_Run2_2017_trackingQuadProp_cff.py` as spurious.

cmsbuild commented:
+1
Tested at: 264cf7bd948a26286bdc81deb226917b9f5c2160
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16911/16946/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16911/16946/summary.html

slava77 commented:
Based on CMSSW_9_0_X_2016-12-10-1100

I can confirm that the technical performance in ttbar pu35 has an overall reduction in CPU: about 10% decrease in tracking itself, which is about 4% of the reco jobs (up to miniaod; no dqm/validation). The non-tracking modules have some moderate increase in CPU, but it's only about 0.5% of the job total. So, the net is about 3.5% reduction.

The file size increases slightly, driven by about 6% in the number of tracks (with a downstream increase 1.8% in PFCandidates). RECO size is up by ~1.3%. I didn't check AOD explicitly, just guessing that it will be ~2-2.5%. This is ~OK by itself.

Looking at https://mkortela.web.cern.ch/mkortela/tracking/validation/CMSSW_9_0_0_pre1_caseeding/ttbar_pu35_ootb.html
probably the worst change is in fake+dups vs dr.
While the efficiency doesn't change much the fake rate is up by a factor of about 3 for btv-like selections (the effect is smaller for inclusive generalTracks).

<img width="409" alt="wf10224_makortel_btvlike_fake_v_dr" src="https://cloud.githubusercontent.com/assets/4676718/21084895/7a01f78a-bfbe-11e6-889a-4866550057c7.png">
The effect is about the same in generalPt09 plot.
Looking in detail: it is primarily from the duplicates
<img width="653" alt="wf10224_makortel_btvlike_fake_dups_v_dr" src="https://cloud.githubusercontent.com/assets/4676718/21085412/9622d22e-bfc6-11e6-8763-d762559cef90.png">

The effect on b-tagging is not very significant: I see the largest effects in high-pt jet sample,
but it varies by the type of the tagger.
in dijet 3 TeV (wf 10059)
<img width="809" alt="wf10059qcd3tev_combmvav2_roc_logl" src="https://cloud.githubusercontent.com/assets/4676718/21085165/cbbd6cea-bfc2-11e6-8846-2f0c8390f33c.png">
<img width="807" alt="wf10059qcd3tev_sisvhe_roc_logl" src="https://cloud.githubusercontent.com/assets/4676718/21085166/d08678e8-bfc2-11e6-9970-4a9b6eb58128.png">

I haven't looked at  plots in my tests yet in detail, just focused on the regions pointed to by the MTV plot vs dr.

Have you noticed this before this PR submission and would consider it is OK?
In the past, we turned down some changes with much smaller impact.
Considering just physics cost and benefits, it seems like this switch brings mainly just costs.
If there is a simple way to reduce duplicates, we better do it with this PR and not introduce a regression to be fixed later.

slava77 commented:
@cmsbuild please test
.. after the spurious file removal.

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: e66db97de074dc1f5da6db528956b18f427ce1dc
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16911/16963/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16911/16963/summary.html

slava77 commented:
@makortel @felicepantaleo 
please clarify if some quick resolution to the excessive duplicate rate is expected (today).
Else, I think we miss pre2.

rovere commented:
Ciao @slava77 ,
thanks for the detailed scrutiny: in fact this increase in the duplicate rate, especially for btv like tracks, somehow sneaked into our internal validation procedure.
Together with @felicepantaleo and @makortel we are actively investigating the issue and trying alternative approaches to reduce the fake rate. Given the current time and the incoming pre2 deadline we see it quite hard for this PR to enter in pre2, but I leave the final word to you guys.
From our (TRK-POG) point of view we would be happy to have this PR integrated as is and provide a fix to the duplicate rate problem later on: this will ease all our development.

Thanks a lot.

slava77 commented:
On 12/12/16 1:25 PM, Marco Rovere wrote:
> Ciao @slava77 <https://github.com/slava77> ,
> thanks for the detailed scrutiny: in fact this increase in the duplicate
> rate, especially for btv like tracks, somehow sneaked into our internal
> validation procedure.
> Together with @felicepantaleo <https://github.com/felicepantaleo> and
> @makortel <https://github.com/makortel> we are actively investigating
> the issue and trying alternative approaches to reduce the fake rate.
> Given the current time and the incoming pre2 deadline we see it quite
> hard for this PR to enter in pre2, but I leave the final word to you guys.
>>From our (TRK-POG) point of view we would be happy to have this PR integrated as is and provide a fix to the duplicate rate problem later on: this will ease all our development.
> 

Ciao Marco,

What is the most important to ease further developments?

I'm guessing it's mostly related to the possible merge conflicts that
can develop quickly based on the set of the files touched.
If that's the case, we can reduce the contention significantly by using
trackingPhase1QuadProp as the default in Run2_2017.

and then the  switch to CA can be done by editing just in ~one place,
to be done once the issue with duplicates is resolved.

		--slava


> Thanks a lot.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/16911#issuecomment-266557403>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcbpdMKZya8npw77Wh4cZG8B5l6JHUks5rHbvNgaJpZM4LH4LI>.
>

rovere commented:
Ciao @slava77 
at this stage your proposal sounds reasonable. I let @makortel work out all the necessary changes on top of this PR.

makortel committed:
Switch the 'old' 2017 tracking back to default, and CA seeding to the special workflows

cmsbuild commented:
Pull request #16911 was updated. @dmitrijus, @cvuosalo, @franzoni, @fabozzi, @cmsbuild, @srimanob, @slava77, @vanbesien, @hengne, @davidlange6 can you please check and sign again.

makortel commented:
@cmsbuild, please test

Following the discussion above, I switched the "old" phase1 seeding back to the default. So essentially this PR now replaced `trackingPhase1PU70` workflows with "CA seeding".

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 8a86cb5394f95f539ae10ca11ccefe1c44252e45
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16911/16997/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16911/16997/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * 20034.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D7_GenSimHLBeamSpotFull14+DigiFull_2023D7+RecoFullGlobal_2023D7+HARVESTFullGlobal_2023D7
   * 23234.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D8_GenSimHLBeamSpotFull14+DigiFull_2023D8+RecoFullGlobal_2023D8+HARVESTFullGlobal_2023D8

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/16911 8a86cb5
- code changes are as described (after rescoping this from a change of the default to just addition of a CA scenario)
- jenkins tests pass and comparisons with baseline show no differences
- some notes on results with CA enabled were posted earlier in the thread before the descope
